### ISSUE TITLE
Non-deterministic Results on valuate DALI loader (HybridValPipe)

### ISSUE THREAD

**Issue by Sakura03** [NVIDIA/DALI#1581]
Hi! I am training ImageNet with DALI dataloader (version 0.13.0). But I get non-deterministic results on multiple runs with exactly the same model architecture and model parameters. The codes are follows:
```
class HybridValPipe(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, data_dir, crop, size):
        super(HybridValPipe, self).__init__(batch_size, num_threads, device_id, seed=12 + device_id)
        if args.use_rec:
            self.input = ops.MXNetReader(path=join(data_dir, "val.rec"), index_path=join(data_dir, "val.idx"),
                                     random_shuffle=False, shard_id=device_id, num_shards=4)
        else:
            self.input = ops.FileReader(file_root=data_dir, shard_id=device_id, num_shards=4, random_shuffle=False)
        self.decode = ops.ImageDecoder(device="mixed", output_type=types.RGB)
        self.res = ops.Resize(device="gpu", resize_shorter=size, interp_type=types.INTERP_TRIANGULAR)
        self.cmnp = ops.CropMirrorNormalize(device="gpu",
                                            output_dtype=types.FLOAT,
                                            output_layout=types.NCHW,
                                            crop=(crop, crop),
                                            image_type=types.RGB,
                                            mean=[0.485 * 255,0.456 * 255,0.406 * 255],
                                            std=[0.229 * 255,0.224 * 255,0.225 * 255])
    def define_graph(self):
        self.jpegs, self.labels = self.input(name="Reader")
        images = self.decode(self.jpegs)
        images = self.res(images)
        output = self.cmnp(images)
        return [output, self.labels]
...
pipe = HybridValPipe(batch_size=64, num_threads=4,
                         device_id=args.local_rank, data_dir=valdir,
                         crop=256, size=224)
pipe.build()
val_loader = DALIClassificationIterator(pipe, size=int(pipe.epoch_size("Reader") / 4))
...
acc1, acc5 = validate(val_loader, model, epoch)
acc1, acc5 = validate(val_loader, model, epoch) # get distinct results between two runs
 ```
The codes are adapted from [https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/mxnet/mxnet-resnet50.html](url).

I have checked the reproducibility of `model`, and same outputs are produced given the same inputs. Besides, I also manually checked that the first batch on the four devices are distinct between two consecutive runs. So I believe the problem comes from my implementation of `HybridValPipe`.

Could anyone please give any suggestions? Many thanks!!!

**Comment by JanuszL** [comment]
Hi,
I don't know what exactly `validate` does in your case. But DALI is an infinite data source. So in you call `validate` again it will pick up the state of DALI when the last call left it.
If `pipe.epoch_size("Reader")` is not divisible equally by the number of GPUs each reader will end up at different sample after the run. If you want to make it 100% reproducible you need to create the pipeline again.

**Comment by zeakey** [comment]
@JanuszL The `validate` function traverses a validation set, e.g. ImageNet, to test the validation accuracy.

After each function call, we reset the val_loader via `val_loader.reset()` so that the cursor should start from the beginning.

**Comment by JanuszL** [comment]
@zeakey - `reset()` resets the iterator state, not the wrapped pipeline.
In your case, when the batch size is 64 and the data set is 50K, then the last batch consists of samples 16 from the first pass over the data set and the reminder - 48 is wrapped up data. When you wrap up then the next call is already 48 ahead in the second pass. To solve this you need to enable `pad_last_batch` in the reader so the last 48 samples will be a padding data (a clone of the last valid sample from the epoch). To avoid passing this rubbish data to the FW use `last_batch_padded` and `fill_last_batch` in the iterator to let it know that this reminder is a dummy data, and it should not be returned to the FW. You can refer to [the MXNet iterator API docs](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/plugins/mxnet_plugin_api.html).
